### PR TITLE
[SYCL-MLIR]: Initialize memory access analysis properly

### PR DESCRIPTION
**Pull request by etiotto**
Address bug exposed in https://github.com/intel/llvm/pull/10298#discussion_r1259919863

### CODE REVIEW

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h [hunk 1]
```diff
@@ -435,24 +435,23 @@ class MemoryAccessAnalysis {
...
+  /// construction.
   MemoryAccessAnalysis &initialize(bool relaxedAliasing) {
-    this->relaxedAliasing = relaxedAliasing;
+    build(relaxedAliasing);
     return *this;
   }
 
   bool isInvalidated(const AnalysisManager::PreservedAnalyses &pa);
 
-  /// Return the operation this analysis was constructed from.
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:445]: Removing: no external users.

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h [hunk 2]
```diff
@@ -496,9 +495,8 @@ class MemoryAccessAnalysis {
   /// The analysis manager.
   AnalysisManager &am;
 
-  /// Whether to assume the program abides to strict aliasing rules (i.e type
```
Comment (etiotto) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:499]: Removing: passed as a argument in `initialized`

File: polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp [hunk 3]
```diff
@@ -1182,12 +1182,11 @@ MemoryAccessPattern MemoryAccess::classify(const MemoryAccessMatrix &matrix,
 //===----------------------------------------------------------------------===//
 
 MemoryAccessAnalysis::MemoryAccessAnalysis(Operation *op, AnalysisManager &am)
-    : operation(op), am(am) {
-  build();
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp:1186]: Now `build` is called via `initialize`.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h [hunk 4]
```diff
@@ -435,24 +435,23 @@ class MemoryAccessAnalysis {
...
-
   /// Return the memory access for the given memref \p access.
   std::optional<MemoryAccess>
   getMemoryAccess(const affine::MemRefAccess &access) const;
 
 private:
   /// Construct the access matrix and offset vector for the memory accesses
   /// contained in the operation associated with the analysis.
-  void build();
+  void build(bool relaxedAliasing);
```
Comment (sommerlukas) [inline polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:454]: Nit: Do we still need both `initialize` and `build`? Seems like the only call to `build` is from `initialize`, so might as well inline that there.

**Review by sommerlukas — APPROVED** [review]

File: polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h (hunk 4 above)
Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h:454]: Yeah initialize may as well build.

Comment (etiotto) [reply polygeist/include/mlir/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.h]: done

**Pull request merged**